### DISCUSSION TITLE
how can i solve this problem? (can't persist nested actions)

### DISCUSSION THREAD

**Discussion by aafnnp** [pmndrs/zustand#1481] [Q&A] [answered]
it's available while first load.but not available when refresh the page

> Environment：next.js 12.2.0 + zustand 4.1.4

```js
import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import shallow from 'zustand/shallow'

interface PersonalInfoState {
	
	actions: {
		queryPersonalInfo: () => void
	}
}

let store: any = (set: any) => ({
	
	actions: {
		queryPersonalInfo: async () => {
			// api here
		},
	},
})

store = devtools(store)
store = persist(store, { name: 'PERSONAL_INFO', getStorage: () => sessionStorage })

export const useInitActions = () => useInitStore(state => state.actions)

export default useInitStore
```
![image](https://user-images.githubusercontent.com/12791216/207767605-3b31fed4-5b5c-41c5-bab1-29ec4ed94858.png)

**Comment by dai-shi** [comment, marked as answer]
Is it reproducible without nextjs, devtools, and/or persist?

**Reply by aafnnp** [reply]
> Is it reproducible without nextjs, devtools, and/or persist?

worked by remove persist,but can not access other state while refresh.

**Reply by dai-shi** [reply]
can you reproduce it with persist without nextjs and devtools?

**Reply by aafnnp** [reply]
> can you reproduce it with persist without nextjs and devtools?

[zustand-demo.zip](https://github.com/pmndrs/zustand/files/10234090/zustand-demo.zip)

**Reply by dai-shi** [reply]
Can you remove nextjs and devtools and see if it repro?
Ideally, make it based on https://codesandbox.io/s/react-typescript-react-ts, only use App.tsx, and make it super minimal?

**Reply by aafnnp** [reply]
> Can you remove nextjs and devtools and see if it repro? Ideally, make it based on https://codesandbox.io/s/react-typescript-react-ts, only use App.tsx, and make it super minimal?

https://codesandbox.io/s/stoic-sun-nch6rc?file=/store/index.ts

**Reply by aafnnp** [reply]
> Can you remove nextjs and devtools and see if it repro? Ideally, make it based on https://codesandbox.io/s/react-typescript-react-ts, only use App.tsx, and make it super minimal?

one more question,Why do I need to delete sessionstoreage and refresh it to get the latest state?

**Reply by dai-shi** [reply]
Thanks for the small reproduction!

The problem is `actions` is serialized as `{}`.
When you restore from session storage, it tries to "merge" states, and `actions: {}` overwrites the defined actions.
To solve this, you have three options:
- do not use actions object
- customize `partialize` option
- customize `merge` option

Here's the option two: https://codesandbox.io/s/jovial-dhawan-uolgud

**Reply by aafnnp** [reply]
> Thanks for the small reproduction!
> 
> The problem is `actions` is serialized as `{}`. When you restore from session storage, it tries to "merge" states, and `actions: {}` overwrites the defined actions. To solve this, you have three options:
> 
> * do not use actions object
> * customize `partialize` option
> * customize `merge` option
> 
> Here's the option two: https://codesandbox.io/s/jovial-dhawan-uolgud

got it

**Reply by acmoles** [reply]
This works fine, but it seems to break Redux Devtools Chrome Extension
